### PR TITLE
OLH-1114 Extract fromUrl from request param or session for Contact Centre Page

### PR DESCRIPTION
**Pull request by DougsCER**
## Proposed changes
`[OLH-1114] Extract fromUrl from request param or session for Contact Centre Page`

### What changed
Check query param and session for a valil fromUrl. If not present then return an error page and log the error. 

### Environment variables or secrets
- [/] No environment variables or secrets were added or changed

## Testing
Unit tests. 
Deployed to dev and saw page loading in all scenarios. Can't yet test the link to contact form because the link will be created on a different ticket.

[OLH-1114]: https://govukverify.atlassian.net/browse/OLH-1114?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/components/contact-govuk-one-login/contact-govuk-one-login-controller.ts [hunk 1]
```diff
@@ -1,7 +1,30 @@
...
-
-  res.render("contact-govuk-one-login/index.njk", data);
+  const fromUrl = getFromUrlAndSaveIt(req);
+  if (fromUrl) {
+    res.render(CONTACT_ONE_LOGIN_TEMPLATE, data);
+  } else {
+    req.log.error(
+      `request and session contain no valid fromUrl`
+    );
+    res.render(NOT_FOUND_TEMPLATE);
```
Comment (DougsCER) [inline src/components/contact-govuk-one-login/contact-govuk-one-login-controller.ts:16]: I'd appreciate your thoughts on the best approach here.

Comment (alex9smith) [reply src/components/contact-govuk-one-login/contact-govuk-one-login-controller.ts:16]: Hmm good point. 

This is just a web page on the internet, so it's entirely possible people end up on it without a URL param (eg. via a Google search). I think we still need to let those people contact us, so if there's no `fromUrl` in the params or in the session then we should not include that param in the link off to the contact form. I'll check with Auth that they can still handle that case.

File: src/components/contact-govuk-one-login/contact-govuk-one-login-controller.ts [hunk 2]
```diff
@@ -1,7 +1,29 @@
...
-export function contactGet(req: Request, res: Response): void {
-  const data = {};
+const CONTACT_ONE_LOGIN_TEMPLATE = "contact-govuk-one-login/index.njk";
+const logger = pino();
 
-  res.render("contact-govuk-one-login/index.njk", data);
+export function contactGet(req: Request, res: Response): void {
+  const fromUrl = getFromUrlAndSaveIt(req);
+  if (!fromUrl) {
+    logger.error(
```
Comment (alex9smith) [inline src/components/contact-govuk-one-login/contact-govuk-one-login-controller.ts:11]:
```suggestion
    logger.info(
```
Now we're not showing users the 404 page when the param isn't present, can you make this INFO level? I don't want to be tripping any alarms when things are working as we expect.

File: src/components/contact-govuk-one-login/contact-govuk-one-login-controller.ts [hunk 3]
```diff
@@ -1,7 +1,29 @@
...
+    );
+  }
+  const data = {
+    fromUrl: fromUrl,
+  };
+  res.render(CONTACT_ONE_LOGIN_TEMPLATE, data);
 }
+
+const getFromUrlAndSaveIt = (request: Request): string => {
+  const fromUrlFromRequest = request.query.fromUrl as string;
```
Comment (alex9smith) [inline src/components/contact-govuk-one-login/contact-govuk-one-login-controller.ts:22]:
```suggestion
  const fromUrlFromRequest = request.query.fromURL as string;
```
We've asked teams to call the parameter `fromURL` (note the capitals). I don't think we need to bother changing the name elsewhere in our code

**Review by alex9smith — CHANGES_REQUESTED** [review]
Just a couple of small changes 👍🏻

File: src/components/contact-govuk-one-login/contact-govuk-one-login-controller.ts (hunk 2 above)
Comment (DougsCER) [reply src/components/contact-govuk-one-login/contact-govuk-one-login-controller.ts:11]: I would argue it's still an error that we are gracefully handling. But unexpected and shouldn't happen. But i can see your point so will update it

File: src/components/contact-govuk-one-login/contact-govuk-one-login-controller.ts [hunk 4]
```diff
@@ -1,7 +1,29 @@
...
-  res.render("contact-govuk-one-login/index.njk", data);
+export function contactGet(req: Request, res: Response): void {
+  const fromURL = getFromUrlAndSaveIt(req);
+  if (!fromURL) {
+    logger.info(
+      "Request to contact-govuk-one-login page did not contain a valid fromURL in the request or session"
+    );
+  }
+  const data = {
+    fromURL: fromURL,
```
Comment (aenciso) [inline src/components/contact-govuk-one-login/contact-govuk-one-login-controller.ts:16]: I think this can be simplified to 
`{ fromURL }`

Comment (DougsCER) [reply src/components/contact-govuk-one-login/contact-govuk-one-login-controller.ts:16]: Resolved this in the next PR 
https://github.com/alphagov/di-account-management-frontend/pull/1058

**Review by alex9smith — APPROVED** [review]

**Pull request merged**